RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1634 @@
+      // Also fill seedRHIdx, topoId, depth
+      for (int rhIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]; rhIdx < nRH;
+           rhIdx += alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]) {
+        int topoId = pfClusteringVars[rhIdx].pfrh_topoId();
+        if (pfClusteringVars[rhIdx].pfrh_isSeed()) {
+          // Valid topo cluster and this rhIdx corresponds to a seed
+          int k = alpaka::atomicAdd(acc, &pfClusteringVars[topoId].rhCount(), 1);
+          int seedIdx = pfClusteringVars[topoId].topoSeedOffsets() + k;
+          if ((unsigned int)seedIdx >= *nSeeds)
+            printf("Warning(contraction) %8d > %8d should not happen, check topoId: %d has %d rh\n",
+                   seedIdx,
+                   *nSeeds,
+                   topoId,
+                   k);
+          pfClusteringVars[seedIdx].topoSeedList() = rhIdx;
+          pfClusteringVars[rhIdx].rhIdxToSeedIdx() = seedIdx;
+          clusterView[seedIdx].topoId() = topoId;
+          clusterView[seedIdx].seedRHIdx() = rhIdx;
+          clusterView[seedIdx].depth() = pfRecHits[rhIdx].depth();
+        }
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      // Determine seed offsets for rechit fraction array
+      for (int rhIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]; rhIdx < nRH;
+           rhIdx += alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]) {
+        pfClusteringVars[rhIdx].rhCount() = 1;  // Reset this counter array
+
+        int topoId = pfClusteringVars[rhIdx].pfrh_topoId();
+        if (pfClusteringVars[rhIdx].pfrh_isSeed() && topoId > -1) {
+          // Allot the total number of rechits for this topo cluster for rh fractions
+          int offset = alpaka::atomicAdd(acc, &totalSeedFracOffset, pfClusteringVars[topoId].topoRHCount());
+
+          // Add offset for this PF cluster seed
+          pfClusteringVars[rhIdx].seedFracOffsets() = offset;
+
+          // Store recHitFraction offset & size information for each seed
+          clusterView[pfClusteringVars[rhIdx].rhIdxToSeedIdx()].rhfracOffset() =
+              pfClusteringVars[rhIdx].seedFracOffsets();
+          clusterView[pfClusteringVars[rhIdx].rhIdxToSeedIdx()].rhfracSize() =
+              pfClusteringVars[topoId].topoRHCount() - pfClusteringVars[topoId].topoSeedCount() + 1;
+        }
+      }
+
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+
+      if (once_per_block(acc)) {
+        pfClusteringVars.pcrhFracSize() = totalSeedFracOffset;
+        pfClusteringVars.nRHFracs() = totalSeedFracOffset;
+        clusterView.nRHFracs() = totalSeedFracOffset;
+        clusterView.nSeeds() = *nSeeds;
+        clusterView.nTopos() = pfClusteringVars.nTopos();
+
+        for (int i = 0; i < nRH; i++) {
+          clusterView[i].topoRHCount() = pfClusteringVars[i].topoRHCount();
+        }
```

Comment:
shouldn't this be in a for loop like
```c++
      for (int rhIdx = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]; rhIdx < nRH;
           rhIdx += alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]) {
           ...
```
outside of `once_per_block` ?

Reply:
Sorry, I think this was made during some development idea that was scrapped. `clusterView` does need a topoRHCount column and this will be removed.